File: formula-py/tests/expected_values.py
from formula import ResolvedValue


class ExpectedValues:
    def __init__(self, yaml: dict):
        self.expected_text = yaml.get("expected_text")
        self.expected_number = yaml.get("expected_number")
        self.expected_boolean = yaml.get("expected_boolean")
        self.expected_list = yaml.get("expected_list")

    def test(self, actual: ResolvedValue):
        if self.expected_text is not None:
            assert actual.as_text() == self.expected_text
        if self.expected_number is not None:
            assert actual.as_number() == self.expected_number
        if self.expected_boolean is not None:
            assert actual.as_boolean() == self.expected_boolean
        if self.expected_list is not None:
            for i in range(len(self.expected_list)):
                ExpectedValues(self.expected_list[i]).test(actual.as_list()[i])

File: java/formula-java/src/test/java/org/formula/test/ExpectedValues.java
package org.formula.test;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.List;
import org.formula.ResolvedValue;

public record ExpectedValues(String expected_text,
                             Integer expected_number,
                             Boolean expected_boolean,
                             List<ExpectedValues> expected_list) {

    public void assertEqualsAll(String key, ResolvedValue actual) {
        if (expected_text != null) {
            assertThat(actual.asText())
                    .as("Expected key '%s' to have text value '%s', but was '%s'", key, expected_text, actual.asText())
                    .isEqualTo(expected_text);
        }
        if (expected_number != null) {
            assertThat(actual.asNumber())
                    .as("Expected key '%s' to have numeric value %d, but was %d", key, expected_number, actual.asNumber())
                    .isEqualTo(expected_number);
        }
        if (expected_boolean != null) {
            assertThat(actual.asBoolean())
                    .as("Expected key '%s' to have boolean value %s, but was %s", key, expected_boolean, actual.asBoolean())
                    .isEqualTo(expected_boolean);
        }
        if (expected_list != null) {
            assertThat(actual.asList()).hasSize(expected_list.size());
            for (int i = 0; i < expected_list.size(); i++) {
                expected_list.get(i).assertEqualsAll(key + "[" + i + "]", actual.asList().get(i));
            }
        }
    }
}
